{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.mode.chained_assignment = None\n",
    "import numpy as np\n",
    "import glob\n",
    "import json\n",
    "import sys,os,argparse\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import plotly\n",
    "plotly.offline.init_notebook_mode(connected=True)\n",
    "import plotly.express as px\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "sns.set(rc={'figure.figsize':(20,8),'axes.facecolor':'white'})\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# command line args\n",
    "CONFIG_FILE = '.config_ipynb'\n",
    "\n",
    "if os.path.isfile(CONFIG_FILE):\n",
    "    with open(CONFIG_FILE) as f:\n",
    "        sys.argv = f.read().split()\n",
    "    \n",
    "parser = argparse.ArgumentParser()\n",
    "parser.add_argument(\"--redsheet\", help=\"Redsheet\")\n",
    "parser.add_argument(\"--batch_path\", help=\"path to QC outputs\")\n",
    "parser.add_argument(\"--xlsx\", help=\"xlsx output file name\")\n",
    "args = parser.parse_args()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Batch QC summary report\n",
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load redsheet into dataframe\n",
    "redsheet = args.redsheet\n",
    "#redsheet = 'test/test_redsheet.csv'\n",
    "manifest = pd.read_csv(f'{redsheet}', skiprows=7, sep=',', index_col='name_alias')\n",
    "\n",
    "# select non Methylseq samples\n",
    "manifest = manifest[(manifest.analysis_type=='Whole Genome Genomic')&~(manifest.eln_sample_id.str.contains('Control'))]\n",
    "\n",
    "# set path for merged BAM QC metrics\n",
    "batch_path = args.batch_path\n",
    "#batch_path = 'test'\n",
    "\n",
    "# open summary metrics file for writing\n",
    "writer =   pd.ExcelWriter(f'{args.xlsx}', engine='xlsxwriter')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def group_layout(fig,df, group, jitter=False):\n",
    "    samples = [len(f['x']) for f in fig['data']]\n",
    "    samples = samples[:df.drop_duplicates(group).shape[0]]\n",
    "    samples = [(i/np.sum(samples)) for i in samples]\n",
    "\n",
    "    ranges = []\n",
    "    col_spacing = 0.005\n",
    "    for i in range(len(samples)):\n",
    "        if i == len(samples):\n",
    "            ranges.append((samples[i-1][-1]+col_spacing),1.0)\n",
    "\n",
    "        elif i == 0:\n",
    "            ranges.append((0.0,samples[i]-col_spacing))\n",
    "        else:\n",
    "            ranges.append((ranges[i-1][-1]+col_spacing,ranges[i-1][-1]+samples[i]-col_spacing))\n",
    "\n",
    "    count = 0\n",
    "    axes = []\n",
    "    for f in fig['layout']:\n",
    "        if 'xaxis' in f:\n",
    "            if f == 'xaxis':\n",
    "                axes.append('xaxis0')\n",
    "            else:\n",
    "                axes.append(f)\n",
    "    axes = sorted(axes, key=lambda x: int(x.split('xaxis')[-1]))\n",
    "    axes[0] = 'xaxis'\n",
    "    for f in axes:\n",
    "        fig.layout[f]['domain'] = ranges[count]\n",
    "        count += 1\n",
    "\n",
    "    for i, a in enumerate(fig.layout.annotations):\n",
    "        r = ranges[i]\n",
    "        fig.layout.annotations[i]['x'] = r[0]+(r[-1]-r[0])/2\n",
    "        fig.layout.annotations[i]['font']['size'] = 12\n",
    "        fig.layout.annotations[i]['text'] = fig.layout.annotations[i]['text']#.replace('Buffer','Buf').replace('Enhancer','Enh').replace('preext','pX')\n",
    "        if jitter:\n",
    "            if i % 2 == 0:\n",
    "                fig.layout.annotations[i]['y'] = fig.layout.annotations[i]['y']+0.05\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def glob_fastp_to_df(manifest,data_path=None, skiprows=None, index_col=None, nrows=None, axis=0, add_sample_name=True, header=0,sep='\\t'):\n",
    "    \n",
    "    input_files = []\n",
    "    input_files += glob.glob(f'{data_path}/*.json')\n",
    "    dfs = []\n",
    "    for file in sorted(input_files):\n",
    "        with open(file) as train_file:\n",
    "            fastp_dict = json.load(train_file)\n",
    "\n",
    "        dict_train = fastp_dict['filtering_result']\n",
    "        sample = file.split('/')[-1].split('.')[0]\n",
    "        df = pd.DataFrame.from_dict(dict_train, orient='index')\n",
    "        df.columns = [sample]\n",
    "        df = df.T\n",
    "        df['lane'] = [l.split('_')[-1] for l in df.index]\n",
    "        df['reads_before_filtering'] = fastp_dict['summary']['before_filtering']['total_reads']\n",
    "        df['sample_name'] = ['_'.join(l.split('_')[:-2]) for l in df.index]\n",
    "        dfs.append(df)\n",
    "    df = pd.concat(dfs)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def glob_to_df(glob_path, skiprows=None, index_col=None, nrows=None, axis=0, col_names=None, add_sample_name=True, normalize=False, header=0,sep='\\t'):\n",
    "    input_files = glob.glob(glob_path)\n",
    "    dfs = []\n",
    "    for input_file in input_files:\n",
    "        sample_name = input_file.split('/')[-1].split('.')[0]\n",
    "        df = pd.read_csv(input_file, sep=sep, skiprows=skiprows, nrows=nrows, index_col=index_col, header=header,names=col_names)\n",
    "        \n",
    "        if normalize != False:\n",
    "            df['percent'] = df[normalize]/df[normalize].sum()*100\n",
    "        if add_sample_name:\n",
    "            df['sample_name'] = '_'.join(sample_name.split('_')[:-1])\n",
    "            df['library_name'] = sample_name\n",
    "        dfs.append(df)\n",
    "    df = pd.concat(dfs, axis=axis)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_df(df, labels=['sample_source','disease_stage','subject_id']):\n",
    "    for label in labels:\n",
    "        df[label] = [manifest.loc[sample_name, label] for sample_name in df.sample_name]\n",
    "    return df\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_fastq_data(df):\n",
    "    dfs = []\n",
    "    for g, group in df.groupby('sample_name'):\n",
    "        group = pd.DataFrame(group.sum(numeric_only=True)).T\n",
    "        for c in group:\n",
    "            group[f'%{c}'] = group[c]/group['reads_before_filtering']*100\n",
    "        group['sample_name'] = g\n",
    "        dfs.append(group)\n",
    "    df = pd.concat(dfs)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## I. fastp QC trimming results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fastp QC trimming results\n",
    "\n",
    "fastq = glob_fastp_to_df(manifest, data_path=f'{batch_path}/fastp')\n",
    "fastq = merge_fastq_data(fastq)\n",
    "fastq = annotate_df(fastq)\n",
    "fastq.sort_values(['sample_source','subject_id'], inplace=True)\n",
    "\n",
    "fastq.to_excel(writer,sheet_name='fastq_filtering')\n",
    "fig = px.bar(fastq, \n",
    "             x='subject_id', \n",
    "             y=['%passed_filter_reads','%low_quality_reads','%too_many_N_reads','%too_short_reads'],\n",
    "             facet_col='sample_source',\n",
    "             title='FASTQ QC summary'\n",
    "            )\n",
    "\n",
    "fig.update_layout(plot_bgcolor='rgba(0,0,0,0)')\n",
    "fig.update_xaxes(showline=True, linewidth=1, linecolor='grey',\n",
    "                 showgrid=False,mirror=True,matches=None,\n",
    "                 tickmode='linear',title='',tickangle=90,\n",
    "                 tickfont=dict(size=10)\n",
    "                )\n",
    "fig.update_yaxes(showline=True, linewidth=0.8, linecolor='grey',\n",
    "                 showgrid=True,mirror=True,gridcolor='lightgrey',gridwidth=0.1)\n",
    "fig.for_each_annotation(lambda a: a.update(text=a.text.split(\"=\")[1],font=dict(size=10)))\n",
    "fig = group_layout(fig,fastq, 'sample_source',jitter=False)\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## II. GC distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalized GC coverage\n",
    "%matplotlib inline\n",
    "\n",
    "gc_data = glob_to_df(f'{batch_path}/picard_CollectMultipleMetrics/*.gc_bias.detail_metrics',\n",
    "                     skiprows=6\n",
    "                    )\n",
    "gc_data = annotate_df(gc_data)\n",
    "\n",
    "\n",
    "groups = gc_data.drop_duplicates('sample_source').shape[0]\n",
    "fig, axs = plt.subplots(groups+2,1,figsize=(15,10),sharex=False,\n",
    "                        gridspec_kw={\"height_ratios\":[1 for i in range(groups+1)]+[0.2]}\n",
    "                       )\n",
    "\n",
    "plt.subplots_adjust(left=0.1,\n",
    "                    bottom=0.1,\n",
    "                    right=0.9,\n",
    "                    top=0.9,\n",
    "                    wspace=0.4,\n",
    "                    hspace=0.01)\n",
    "axs = axs.flatten()\n",
    "axs[0] = sns.lineplot(x='GC', y='NORMALIZED_COVERAGE', hue='sample_source', data=gc_data[np.abs(gc_data['NORMALIZED_COVERAGE']-gc_data['NORMALIZED_COVERAGE'].mean()) <= (0.2*gc_data['NORMALIZED_COVERAGE'].std())].reset_index(), ci=None, ax=axs[0])\n",
    "axs[0].hlines(1,0,100,color='black',linestyles='dashed')\n",
    "axs[0].set_ylim(0,2)\n",
    "axs[0].set_xlim(0,95)\n",
    "axs[0].legend(loc = \"upper right\")\n",
    "axs[0].set_title('Mean Normalized GC coverage',fontsize=16)\n",
    "\n",
    "i = 1\n",
    "dfs = []\n",
    "for g, group in gc_data.groupby('sample_source'):\n",
    "    y_values = []\n",
    "    samples_names = []\n",
    "    for s, sample in group.groupby('subject_id'):\n",
    "        sample = sample[np.abs(sample['NORMALIZED_COVERAGE']-sample['NORMALIZED_COVERAGE'].mean()) <= (0.5*sample['NORMALIZED_COVERAGE'].std())]\n",
    "        y = np.array(sample.NORMALIZED_COVERAGE.tolist())\n",
    "        y_values.append(y)\n",
    "        samples_names.append(s)\n",
    "    matrix = np.matrix(y_values)\n",
    "    matrix = pd.DataFrame(matrix)\n",
    "    matrix['sample_name'] = samples_names\n",
    "    matrix.set_index('sample_name', inplace=True)\n",
    "    axs[i] = sns.heatmap(matrix,ax=axs[i],vmax=2, cbar=False)\n",
    "    axs[i].set_yticklabels('')\n",
    "    axs[i].set_ylabel(g)\n",
    "    axs[i].set_xlim(0,95)\n",
    "    axs[i].sharex(axs[0])\n",
    "    i+=1\n",
    "    dfs.append(matrix)\n",
    "\n",
    "x = axs[-1].set_xlabel('% GC')\n",
    "c = fig.colorbar(axs[-1].imshow(matrix,vmax=2), cax=axs[-1], orientation=\"horizontal\",anchor=(0.9, 1.0))\n",
    "\n",
    "matrix = pd.concat(dfs)\n",
    "matrix = matrix.T\n",
    "matrix['%GC'] = matrix.index.tolist()\n",
    "matrix.to_excel(writer,sheet_name='GC distribution')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## III. AT and GC drop out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# AT drop out\n",
    "gc = glob_to_df(f'{batch_path}/picard_CollectMultipleMetrics/*gc_bias.summary_metrics',skiprows=6)\n",
    "gc = annotate_df(gc)\n",
    "gc.sort_values(['disease_stage','subject_id'], inplace=True)\n",
    "gc.to_excel(writer,sheet_name='GC_bias')\n",
    "fig = px.bar(gc, \n",
    "             x='subject_id', \n",
    "             y='AT_DROPOUT',\n",
    "             color='sample_source',\n",
    "             facet_col='sample_source',\n",
    "             title='AT DROPOUT',\n",
    "             height=350\n",
    "            )\n",
    "\n",
    "fig.update_layout(plot_bgcolor='rgba(0,0,0,0)')\n",
    "fig.update_xaxes(showline=True, linewidth=1, linecolor='grey',\n",
    "                 showgrid=False,mirror=True,matches=None,\n",
    "                 tickmode='linear',title='',tickangle=90,\n",
    "                 tickfont=dict(size=10)\n",
    "                )\n",
    "fig.update_yaxes(showline=True, linewidth=0.8, linecolor='grey',\n",
    "                 showgrid=True,mirror=True,gridcolor='lightgrey',gridwidth=0.1)\n",
    "fig.for_each_annotation(lambda a: a.update(text=a.text.split(\"=\")[1],font=dict(size=10)))\n",
    "fig = group_layout(fig,gc, 'sample_source',jitter=False)\n",
    "\n",
    "fig.show()\n",
    "\n",
    "fig = px.bar(gc, \n",
    "             x='subject_id', \n",
    "             y='GC_DROPOUT',\n",
    "             color='sample_source',\n",
    "             facet_col='sample_source',\n",
    "             title='GC DROPOUT',\n",
    "             height=350\n",
    "            )\n",
    "\n",
    "fig.update_layout(plot_bgcolor='rgba(0,0,0,0)')\n",
    "fig.update_xaxes(showline=True, linewidth=1, linecolor='grey',\n",
    "                 showgrid=False,mirror=True,matches=None,\n",
    "                 tickmode='linear',title='',tickangle=90,\n",
    "                 tickfont=dict(size=10)\n",
    "                )\n",
    "fig.update_yaxes(showline=True, linewidth=0.8, linecolor='grey',\n",
    "                 showgrid=True,mirror=True,gridcolor='lightgrey',gridwidth=0.1)\n",
    "fig.for_each_annotation(lambda a: a.update(text=a.text.split(\"=\")[1],font=dict(size=10)))\n",
    "fig = group_layout(fig,fastq, 'sample_source',jitter=False)\n",
    "\n",
    "fig.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## IV. Insert size distributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert size distributions\n",
    "insert_size = glob_to_df(f'{batch_path}/picard_CollectInsertSizeMetrics/*.insert_size_metrics.txt',\n",
    "                     skiprows=11,normalize='All_Reads.fr_count'\n",
    "                    )\n",
    "insert_size = annotate_df(insert_size)\n",
    "\n",
    "insert_size.sort_values(['disease_stage','subject_id'], inplace=True)\n",
    "\n",
    "insert_size.to_excel(writer, sheet_name='fragment_lengths')\n",
    "\n",
    "\n",
    "fig = px.line(insert_size, x='insert_size', y='percent', color='sample_source',hover_name='sample_source',\n",
    "              line_group='subject_id',\n",
    "              title='Mean fragment length distribution per sample type',\n",
    "             )\n",
    "fig.update_layout(plot_bgcolor='rgba(0,0,0,0)')\n",
    "fig.update_xaxes(showline=True, linewidth=1, linecolor='grey',title='fragment length',\n",
    "                 showgrid=False,mirror=True,matches=None)\n",
    "fig.update_yaxes(showline=True, linewidth=0.8, linecolor='grey',title='percent',\n",
    "                 showgrid=False,mirror=True,gridcolor='lightgrey',gridwidth=0.1)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "insert_size = glob_to_df(f'{batch_path}/picard_CollectInsertSizeMetrics/*.insert_size_metrics.txt',\n",
    "                     skiprows=6,nrows=1)\n",
    "insert_size = annotate_df(insert_size)\n",
    "\n",
    "\n",
    "insert_size.sort_values(['disease_stage','subject_id'], inplace=True)\n",
    "\n",
    "insert_size.to_excel(writer, sheet_name='fragment_length_stats')\n",
    "\n",
    "fig = px.bar(insert_size, x='subject_id', y='MEDIAN_INSERT_SIZE',facet_col='sample_source',color='sample_source',\n",
    "             title='Median fragment length per sample',height=350)\n",
    "fig.update_layout(plot_bgcolor='rgba(0,0,0,0)')\n",
    "fig.update_xaxes(showline=True, linewidth=1, linecolor='grey',\n",
    "                 showgrid=False,mirror=True,matches=None,\n",
    "                 tickmode='linear',title='',tickangle=90,\n",
    "                 tickfont=dict(size=10)\n",
    "                )\n",
    "fig.update_yaxes(showline=True, linewidth=0.8, linecolor='grey',range=(100,200),\n",
    "                 showgrid=True,mirror=True,gridcolor='lightgrey',gridwidth=0.1)\n",
    "fig.for_each_annotation(lambda a: a.update(text=a.text.split(\"=\")[1],font=dict(size=10)))\n",
    "fig = group_layout(fig,insert_size, 'sample_source',jitter=False)\n",
    "fig.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## V. Coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "depth = glob_to_df(f'{batch_path}/mosdepth/*.mosdepth.summary.txt')\n",
    "depth = depth[depth.chrom=='total']\n",
    "depth = annotate_df(depth)\n",
    "depth.sort_values(['sample_source','subject_id'], inplace=True)\n",
    "depth.to_excel(writer, sheet_name='depth_stats')\n",
    "fig = px.bar(depth, x='subject_id', y='mean',facet_col='sample_source',\n",
    "             color='sample_source',title='Mean coverage across the human genome',\n",
    "             height=350\n",
    "            )\n",
    "fig.update_layout(plot_bgcolor='rgba(0,0,0,0)')\n",
    "fig.update_xaxes(showline=True, linewidth=1, linecolor='grey',\n",
    "                 showgrid=False,mirror=True,matches=None,\n",
    "                 tickmode='linear',title='',tickangle=90,\n",
    "                 tickfont=dict(size=10)\n",
    "                )\n",
    "fig.update_yaxes(showline=True, linewidth=0.8, linecolor='grey',\n",
    "                 showgrid=True,mirror=True,gridcolor='lightgrey',gridwidth=0.1)\n",
    "fig.for_each_annotation(lambda a: a.update(text=a.text.split(\"=\")[1],font=dict(size=10)))\n",
    "fig = group_layout(fig,depth, 'sample_source',jitter=False)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "depth = glob_to_df(f'{batch_path}/mosdepth/*.mosdepth.global.dist.txt',header=None, col_names=['chrom','cov','percent'])\n",
    "depth = depth[depth.chrom=='total']\n",
    "depth = depth[depth.percent>0]\n",
    "dfs = []\n",
    "for g, group in depth.groupby('sample_name'):\n",
    "    sample_source = manifest.loc[g,'sample_source']\n",
    "    disease_stage = manifest.loc[g,'disease_stage']\n",
    "    subject_id = manifest.loc[g,'subject_id']\n",
    "    group['sample_source'] = sample_source\n",
    "    group['disease_stage'] = disease_stage\n",
    "    group['subject_id'] = subject_id\n",
    "    dfs.append(group)\n",
    "depth = pd.concat(dfs)\n",
    "depth.sort_values(['disease_stage','subject_id'], inplace=True)\n",
    "\n",
    "\n",
    "\n",
    "fig = px.line(depth, x='cov', y='percent', color='sample_source',hover_name='sample_source',\n",
    "              line_group='subject_id',\n",
    "              title='Whole genome coverage distribution',\n",
    "             )\n",
    "fig.update_layout(plot_bgcolor='rgba(0,0,0,0)')\n",
    "fig.update_xaxes(showline=True, linewidth=1, linecolor='grey',title='fragment length',\n",
    "                 showgrid=False,mirror=True,matches=None)\n",
    "fig.update_yaxes(showline=True, linewidth=0.8, linecolor='grey',title='percent',\n",
    "                 showgrid=False,mirror=True,gridcolor='lightgrey',gridwidth=0.1)\n",
    "fig.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## VI. Mapping QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# picard metrics\n",
    "picard = glob_to_df(f'{batch_path}/picard_CollectMultipleMetrics/*alignment_summary_metrics',skiprows=6,nrows=3)\n",
    "picard = picard[picard.CATEGORY == 'PAIR']\n",
    "picard = annotate_df(picard)\n",
    "picard.sort_values(['sample_source','subject_id'], inplace=True)\n",
    "\n",
    "picard.to_excel(writer, sheet_name='picard_metrics')\n",
    "\n",
    "metrics = ['PCT_PF_READS_ALIGNED',\n",
    "           'PCT_PF_READS_IMPROPER_PAIRS',\n",
    "           'PCT_CHIMERAS',\n",
    "           'PCT_SOFTCLIP',\n",
    "           'PF_HQ_ERROR_RATE',\n",
    "           'PF_INDEL_RATE',\n",
    "           'PF_ALIGNED_BASES'\n",
    "          ]\n",
    "for metric in metrics:\n",
    "        fig = px.bar(picard,x='subject_id',y=metric,\n",
    "                     labels = dict(y='skjskdj'),\n",
    "                     template='plotly',\n",
    "                     height=350,color='sample_source',width=990,\n",
    "                     facet_col='sample_source'\n",
    "                    )\n",
    "\n",
    "\n",
    "\n",
    "        fig.update_layout(plot_bgcolor='rgba(0,0,0,0)',showlegend=False)\n",
    "        fig.update_xaxes(showline=True, linewidth=1, linecolor='grey',\n",
    "                         showgrid=False,mirror=True,title='',tickmode='linear',matches=None,\n",
    "                        tickangle=90)\n",
    "        fig.update_yaxes(showline=True, linewidth=0.8, linecolor='grey',title_font=dict(size=11),\n",
    "                         showgrid=False,mirror=True,gridcolor='lightgrey',gridwidth=0.1)\n",
    "\n",
    "        fig.for_each_annotation(lambda a: a.update(text=a.text.split(\"=\")[1],font=dict(size=10)))\n",
    "        fig = group_layout(fig,picard, 'sample_source')\n",
    "        fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "writer.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
